src/cargo/util/toml/mod.rs

```diff
@@ -419,18 +420,71 @@ impl<'de> de::Deserialize<'de> for ProfilePackageSpec {
 }
 
 impl TomlProfile {
-    fn validate(
+    pub fn from_config(
+        config: &Config,
+        name: &str,
+        warnings: &mut Vec<String>,
+    ) -> CargoResult<Option<TomlProfile>> {
+        if !config.cli_unstable().config_profile {
+            return Ok(None);
+        }
+        if let Some(util::config::Value { val, .. }) =
+            config.get_table(&format!("profile.{}", name))?
```

Comment:
One of the downsides of `get_table` is that it [doesn't work well with environment variables](https://github.com/rust-lang/cargo/issues/5416) so I think that for any new features we ideally want to avoid `get_table` as much as possible. I saw above that `get_table` is used a lot in validation which I think is fine, but would it be possible for these values to read from the environment as well?

Reply:
I thought about that a little bit.  IIUC, it will need to manually `get` every key, and it wouldn't be able to use `serde` to do the automatic conversion.   I can give it a shot, though I think that will make the patch a little longer.

(Another crazy idea would be to implement a custom Deserializer for Config that would handle everything, but I'm not sure if that would work, and would be a fair amount of code.)

Reply:
Heh yeah I think you're spot on. The manual deserialization here hopefully isn't too bad, but I agree that a `Config` deserializer is probably what we want long-term.